core/src/storage2/lazy/lazy_cell.rs

```diff
@@ -325,12 +324,13 @@ where
         // SAFETY: This is critical because we mutably access the entry.
         let cache = unsafe { &mut *self.cache.get_ptr().as_ptr() };
         if cache.is_none() {
-            // Write value to storage
-            let key = self
-                .key
-                .expect("key must exist when attempting to set a value");
-            push_spread_root_opt::<T>(Some(&new_value), &key);
+            // Cache is empty, so we simply set the cache to the value.
+            // The key does not need to exist for this to work, we only need to
+            // write the value into the cache and are done. Writing to contract
+            // storage happens during setup/teardown of a contract.
+            *cache = Some(Entry::new(Some(new_value), EntryState::Preserved));
```

Comment:
```suggestion
            *cache = Some(Entry::new(Some(new_value), EntryState::Mutated));
```

Reply:
The `EntryState` must be `Mutated`, otherwise the `Entry`'s value is not going to be written back on contract tear-down. `Preserved` means that the cache holds a value that mirror what can already be found in the contract storage.